clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java

```diff
@@ -59,136 +60,140 @@
-            // We partially construct FetchablePartitionResponse since we don't know the partition ID at this point
-            // When we convert the PartitionData (and other fields) into FetchResponseData down in toMessage, we
-            // set the partition IDs.
-            this.partitionResponse = partitionResponse;
-            this.preferredReplica = Optional.of(partitionResponse.preferredReadReplica())
-                .filter(replicaId -> replicaId != INVALID_PREFERRED_REPLICA_ID);
-
-            if (partitionResponse.abortedTransactions() == null) {
-                this.abortedTransactions = null;
-            } else {
-                this.abortedTransactions = partitionResponse.abortedTransactions().stream()
-                    .map(AbortedTransaction::fromMessage)
-                    .collect(Collectors.toList());
-            }
-
-            this.error = Errors.forCode(partitionResponse.errorCode());
-        }
+    @Override
+    public boolean shouldClientThrottle(short version) {
+        return version >= 8;
+    }
 
-        public PartitionData(Errors error,
-                             long highWatermark,
-                             long lastStableOffset,
-                             long logStartOffset,
-                             Optional<Integer> preferredReadReplica,
-                             List<AbortedTransaction> abortedTransactions,
-                             Optional<FetchResponseData.EpochEndOffset> divergingEpoch,
-                             T records) {
-            this.preferredReplica = preferredReadReplica;
-            this.abortedTransactions = abortedTransactions;
-            this.error = error;
-
-            FetchResponseData.FetchablePartitionResponse partitionResponse =
-                new FetchResponseData.FetchablePartitionResponse();
-            partitionResponse.setErrorCode(error.code())
-                .setHighWatermark(highWatermark)
-                .setLastStableOffset(lastStableOffset)
-                .setLogStartOffset(logStartOffset);
-            if (abortedTransactions != null) {
-                partitionResponse.setAbortedTransactions(abortedTransactions.stream().map(
-                    aborted -> new FetchResponseData.AbortedTransaction()
-                        .setProducerId(aborted.producerId)
-                        .setFirstOffset(aborted.firstOffset))
-                    .collect(Collectors.toList()));
-            } else {
-                partitionResponse.setAbortedTransactions(null);
-            }
-            partitionResponse.setPreferredReadReplica(preferredReadReplica.orElse(INVALID_PREFERRED_REPLICA_ID));
-            partitionResponse.setRecordSet(records);
-            divergingEpoch.ifPresent(partitionResponse::setDivergingEpoch);
 
-            this.partitionResponse = partitionResponse;
-        }
+    public static final class PartitionData<T extends BaseRecords> {
```

Comment:
Do we need this class? What does it add over `FetchResponseData.FetchablePartitionRespons`?

Reply:
> What does it add over FetchResponseData.FetchablePartitionRespons?

It offers methods to transfer "non-defined value" to ```Optional``` type.

Reply:
It seems like we do it for a couple of cases. It doesn't seem worth it given that fetch is one of the hot paths. We could have static helper methods instead if helpful.

Reply:
Nice suggestions. will address it.